### ISSUE TITLE
[question] How hard would it be to implement 4-bit precision training?

### ISSUE THREAD

**Issue by michalsustr** [pytorch/pytorch#49298]
I came across the paper [Ultra-Low Precision 4-bit Training of Deep Neural Networks](https://proceedings.neurips.cc/paper/2020/file/13b919438259814cd5be8cb45877d577-Paper.pdf) on NeurIPS 2020. I think it would be cool to implement support for it in PyTorch. I think it can be done quite efficiently on CPU using the AVX2 instruction set, as all the multiplication/addition operations can be stored in a fast cache. The operations would just make a lookup in this table.

I had a look how things are implemented in the library. If I am correct, there is enough of level of abstraction to make this doable. I need to implement a kernel and add it to the ATEN's DispatchStub or something like that. If I copy-paste implementation of `
pytorch/aten/src/ATen/quantized/` and make it work with custom `fp4` type that should work for end-to-end training, right? To start playing around with this, like training my own MNIST, it should be enough to just implement addition and multiplication for something like MLP with relus: all computation consists only from  affine operations so + and * should be enough. 

I would appreciate high-level guidance / help links on this. Thank you!

**Comment by glaringlee** [comment]
cc @ngimel @VitalyFedyunin

**Comment by ezyang** [comment]
4-bit would be nifty. It has some similar problems to https://github.com/pytorch/pytorch/issues/32867 where 4bit is smaller than the addressable 8bit so the framework would have to be extended in some pretty fundamental ways to make this work.

**Comment by ngimel** [comment]
I think quantization already has some sort of support for 4-bit.

**Comment by vadimkantorov** [comment]
@ezyang Maybe it would be fine using some framework like Vec<T>? e.g. only operate with vector instructions (some padding may be needed, which may violate view/reshape semantics) at least for the case when padding is not needed

**Comment by ezyang** [comment]
Yeah, I think vectorized types are probably the right way to do this.
